### DISCUSSION TITLE
Intercept Access Token  Authorization Code grant

### DISCUSSION THREAD

**Discussion by pommelinho** [ory/hydra#2364] [Q&A]
Hey!
I am using hydra along with my own implementation of Login/Consent Flow to implement OAuth 2.0   Authorization Code grant.
When user provided consent the access_token and refresh_token will be sent to the redirect_url of the client.
Is is somehow possible that i can get the  access_token and refresh_token from hydra before it is sent to the client?

Background: I want to store some device info of mobile devices along with the provided access_token, refresh_token in my database.

**Comment by aeneasr** [comment]
> Is is somehow possible that i can get the access_token and refresh_token from hydra before it is sent to the client?

No there is not!

> Background: I want to store some device info of mobile devices along with the provided access_token, refresh_token in my database.

Don't do that! We have so many safety features in place to avoid exactly this - storing plain text credentials in the database. This is really a bad practice - like storing a password in plain text in your database.

**Reply by pommelinho** [reply]
@aeneasr  but i am not storing any plain text credentials in the database?
 so you are saying that i should not store the access_tokens  in my database to connect them to devices which are using them. hence for the sake of security there is no possibility to revoke access for a specific device as i do not know which token this device is using, right?

**Reply by vinckr** [reply]
Could you use the `client_id` for this purpose instead?

**Reply by pommelinho** [reply]
The client in my case is the mobile application itself. So i can revoke all the tokens for the mobile application but not for a specific device. Or am i missunderstanding your idea?

**Reply by vinckr** [reply]
Just to be clear, 
you want to revoke the Access token and not just invalidate the user session, is that correct?

Maybe this can help you out: [Lists All Consent Sessions of a Subject](https://www.ory.sh/hydra/docs/reference/api/#lists-all-consent-sessions-of-a-subject)
I believe you can save any info that you want to add in `metadata`.

**Reply by pommelinho** [reply]
Exactly i want to revoke access and refresh token. But not  all tokens for a specific subject  e.g a user email but **only one token** for a **e.g. a specific mobile device** of that subject.
So that that the tokens issued to device A  for that subject are still valid but the tokens issued for device B for that subject not.
But as far as i understand it it is only possible to revoke **all tokens for one subject** by using  this endpoint https://www.ory.sh/hydra/docs/reference/api/#revokes-consent-sessions-of-a-subject-for-a-specific-oauth-20-client

One idea from my side would be to store in the subject beside the email also the deviceId , formate email_deviceId (e.g. test@test.de_fa4fcb8c54ca9254)

**Reply by vinckr** [reply]
You could also store arbitrary data in `metadata`. 

Could you give me an example of an use case where you would need to revoke access/refresh tokens for a specific device of a subject? 

It sounds to me as though you are talking about sessions, probably some misunderstanding!

**Reply by vinckr** [reply]
By the way is this related to your other question: https://github.com/ory/hydra/discussions/2365 ? 
I couldnt quite figure it out...

**Reply by pommelinho** [reply]
> You could also store arbitrary data in `metadata`.
> 
> Could you give me an example of an use case where you would need to revoke access/refresh tokens for a specific device of a subject?
> 
> It sounds to me as though you are talking about sessions, probably some misunderstanding!

An Example would be that you the user (the subject) can access a settings page, where he sees his valid sessions associated to a specific mobile device or browser. Now he want to invalidates the session only for Device A  as this was stolen, but keep it for Device B.
However https://www.ory.sh/hydra/docs/reference/api/#revokes-consent-sessions-of-a-subject-for-a-specific-oauth-20-client only allows me to revoke all consent sessions for one subject.

**Reply by pommelinho** [reply]
> By the way is this related to your other question: #2365 ?
> I couldnt quite figure it out...

No not really.  Depends on the solution of my problem here ;-)

**Reply by vinckr** [reply]
I think I found a solution,
use `challenge`
> ID is the identifier ("authorization challenge") of the consent authorization request. It is used to identify the session.

Use this endpoint to get all the sessions, also the challenges. [Lists All Consent Sessions of a Subject](https://www.ory.sh/hydra/docs/reference/api#lists-all-consent-sessions-of-a-subject)
You can save something related to the device in `metadata` and then use `challenge` to identify the consent session you want to terminate. 
You need to do this in your consent app, there is [a reference implementation](https://github.com/ory/hydra-login-consent-node) here.

Let me know if that helps you, I have not tested it but it seems this is a viable solution.

> related issue https://github.com/ory/hydra/issues/902

**Reply by pommelinho** [reply]
Hey first thank you for your respons.
Two questions are arising here.

1. Based on the challenge, how can i terminate the consent session. https://www.ory.sh/hydra/docs/reference/api/#revokes-consent-sessions-of-a-subject-for-a-specific-oauth-20-client endpoint only allows this by subject as far as I understand

2.According to https://github.com/ory/hydra-login-consent-node/blob/master/src/routes/consent.ts: Where can i pass the metadata here  for s specific consent request? As far as I understand the` metadata` when[ listing all consent sessions of a subjec](https://www.ory.sh/hydra/docs/reference/api/#lists-all-consent-sessions-of-a-subject)t is always same for a specific client?

**Reply by vinckr** [reply]
1. Yes you need to associate the subject with the challenge somehow. At least that would be my idea how to solve this.

2. True, I think you are right there. Thats why I would recommend doing this in your IAM or Kratos.
This issue is related to what you are trying to do: https://github.com/ory/kratos/issues/397

**Reply by pommelinho** [reply]
Ok thank you.

Regarding 1.) 
So how would you associate it?  In my eyes we would need to add a suffix (e.g. the challeng) to the subject  or something like this...

**Reply by vinckr** [reply]
This could also be useful: https://www.ory.sh/hydra/docs/next/reference/api/#invalidates-all-login-sessions-of-a-certain-user

I think using the challenge as identifier would work yes!

**Reply by pommelinho** [reply]
Ok, thank you for your help!
